### DISCUSSION TITLE
[HW1] problem 3.4

### DISCUSSION THREAD

**Discussion by emjay73** [seominjoon/kaist-ai605#78] [Q&A] [answered]
According to the article below, CrossEntropyLoss supported by PyTorch itself applies log softmax to the input.
https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html.
So the following statements in the assignment confuse me.

```
note that we use logits instead of softmax(logits) as the input to the cross entropy, which allow us to avoid numerical instability.
```
```
Problem 3.4 (bonus) (1 points) Why is it numerically unstable if you compute log on top of softmax?
```

Do I have to elaborate on why we should not use log softmax twice?

**Comment by seominjoon** [comment, marked as answer]
No, you should not use log softmax twice.
As you said, PyTorch CrossEntropyLoss applies log(softmax()) instead of just softmax(). The question is asking how this is helpful for reducing numerical instability.

**Reply by seominjoon** [reply]
In other words: an alternative choice (by PyTorch) would have been that you apply log to the output of softmax yourself. PyTorch doesn't want you to do this though. Why would that be?

**Reply by emjay73** [reply]
So rephrasing the question would be,
"Why adopting Softmax and Log in a single expression is more preferable compared to the 'softmax and then log'?"
Is that correct?

**Reply by seominjoon** [reply]
@emjay73 That's correct!

**Reply by emjay73** [reply]
Thank you!
